giftless/storage/aws_s3.py

```diff
@@ -0,0 +1,110 @@
+
+from giftless.storage import ExternalStorage, StreamingStorage
+from giftless.storage.exc import ObjectNotFound
+
+
+class AwsS3Storage(StreamingStorage, ExternalStorage):
+    """AWS S3 Blob Storage backend.
+    """
+
+    def __init__(self, aws_s3_bucket_name: str, path_prefix: Optional[str] = None, **_):
+        self.aws_s3_bucket_name = aws_s3_bucket_name
+        self.path_prefix = path_prefix
+        self.s3 = boto3.resource('s3')
+        self.s3_client = boto3.client('s3')
+
+    def get(self, prefix: str, oid: str) -> Iterable[bytes]:
+        if not self.exists(prefix, oid):
+            raise ObjectNotFound()
+        result: Iterable[bytes] = self._s3_object(prefix, oid).get()['Body']
+        return result
+
+    def put(self, prefix: str, oid: str, data_stream: BinaryIO) -> int:
+        completed = []
+
+        def upload_callback(size):
+            completed.append(size)
+
+        bucket = self.s3.Bucket(self.aws_s3_bucket_name)
+        bucket.upload_fileobj(data_stream, self._get_blob_path(prefix, oid), Callback=upload_callback)
+        return sum(completed)
+
+    def exists(self, prefix: str, oid: str) -> bool:
+        s3_object = self._s3_object(prefix, oid)
+        try:
+            s3_object.load()
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                return False
+            else:
+                raise RuntimeError(e)
+        return True
+
+    def get_size(self, prefix: str, oid: str) -> int:
+        if self.exists(prefix, oid):
+            result: int = self._s3_object(prefix, oid).content_length
+            return result
+        else:
+            raise ObjectNotFound()
+
+    def get_upload_action(self, prefix: str, oid: str, size: int, expires_in: int,
+                          extra: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
+        params_ = {
+            'Bucket': self.aws_s3_bucket_name,
+            'Key': self._get_blob_path(prefix, oid)
+        }
+        response = self.s3_client.generate_presigned_url('put_object',
+                                                         Params=params_,
+                                                         ExpiresIn=expires_in
+                                                         )
+        return {
+            "actions": {
+                "upload": {
+                    "href": response,
+                    "header": {},
+                    "expires_in": expires_in
+                }
+            }
+        }
+
+    def get_download_action(self, prefix: str, oid: str, size: int, expires_in: int,
+                            extra: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
+
+        filename = extra.get('filename') if extra else oid
+        params_ = {
+            'Bucket': self.aws_s3_bucket_name,
+            'Key': self._get_blob_path(prefix, oid),
+            'ResponseContentDisposition': f"attachment; filename = {filename}"
```

Comment:
There shouldn't be any spaces around the `=` in the header, I'm not sure it is valid like that. Also, you may need to quote and escape or at least sanitize the `filename` variable before putting it in a header. See the use of `safe_filename` in https://github.com/datopian/giftless/blob/2f1432c4bb679fdbc66bda34ce841d293dc02834/giftless/transfer/basic_streaming.py#L85 as example. Also, see https://github.com/datopian/giftless/issues/70 complaining about the same function and how it is too strict ;)